src/clj_3df/compiler.cljc

```diff
@@ -496,12 +505,16 @@
 (defmethod unify [Projection Predicate] [proj pred] (update proj :binding unify pred))
 ;; @TODO compose predicate functions
 (defmethod unify [Predicate Predicate] [p1 p2] (update p1 :binding unify p2))
+(defmethod unify [Predicate FnExpr] [pred f] (unify f pred))
 
 (defmethod unify [::binding FnExpr] [b1 b2] (unify b2 b1))
 (defmethod unify [FnExpr ::binding] [f binding] (update f :binding unify binding))
 (defmethod unify [FnExpr Projection] [b1 b2] (unify b2 b1))
 (defmethod unify [Projection FnExpr] [proj f] (update proj :binding unify f))
 (defmethod unify [FnExpr FnExpr] [p1 p2] (update p1 :binding unify p2))
+(defmethod unify [FnExpr Predicate] [f pred] (update f :binding unify pred))
```

Comment:
I think we should throw here for now. Should never happen in practice, as long as output bindings are distinct from input bindings for function expressions.

Reply:
Fair!